eurito_daps/packages/cordis/cordis_neo4j.py

```diff
@@ -0,0 +1,246 @@
+
+    Args:
+        orm_instance (sqlalchemy.Base): Instance of a SqlAlchemy ORM, i.e.
+                                        a 'row' of data.
+    Returns:
+        row (dict): A flat row of data, inferred from `orm_instance`
+    """
+    row = object_to_dict(orm_instance, shallow=True)
+    return {k: _flatten(v) for k, v in row.items()}
+
+
+def _flatten(data):
+    """Flatten JSON to a flat dictionary via a hard-coded routine.
+
+    Args:
+        data (json-like): Input data.
+    Returns:
+        _data (dict): A flat dictionary.
+    """
+    # Flatten dictionaries directly
+    if type(data) is dict:
+        return flatten_dict(data)
+    # if not a dict or a list, just return unaltered
+    elif type(data) is not list:
+        return data
+    # Assume it is now a list. Check item types to determine next step.
+    types = set(type(row) for row in data)
+    if len(types) > 1:
+        raise TypeError(f'Mixed types ({types}) are not accepted')
+    # Flatten the internal data if the item types are dict
+    try:
+        if next(iter(types)) is dict:
+            return [flatten_dict(row) for row in data]
+    except StopIteration:  # Implies empty list
+        pass
+    # Otherwise don't flatten
+    return data
+
+
+def flatten_dict(row, keys=[('title',),
+                            ('street', 'city', 'postalCode')]):
+    """Flatten a dict by concatenating string values
+    of matching keys.
+
+    Args:
+        row (dict): Data to be flattened
+    Returns:
+        flat (str): Concatenated data.
+    """
+    flat = ''  # The output data
+    for ks in keys:
+        # If any keys are present, join the values
+        if not any(k in row for k in ks):
+            continue
+        flat = '\n'.join(row[k] for k in ks
+                         if k in row)
+        break    
+    assert len(flat) > 0  # Ensures that a key has been found, 
+                          # otherwise you'll need to provide more keys
+    return flat
+
+
+def retrieve_node(session, graph, orm, parent_orm, data_row):
+    """Retrieve an existing node from neo4j, by first retrieving it's
+    id (field name AND value) via SqlAlchemy.
+
+    Args:
+        session (sqlalchemy.Session): SQL DB session.
+        transaction (py2neo.Transaction): Neo4j transaction
+        orm (sqlalchemy.Base): SqlAlchemy ORM describing :obj:`data_row`
+        parent_orm (sqlalchemy.Base): Parent ORM to build relationship to
+        data_row (dict): Flat row of data retrieved from `orm`
+    Returns:
+        node (py2neo.Node): Node of data corresponding to data_row
+    """
+    row = get_row(session, parent_orm, orm, data_row)
+    (pk,) = inspect(parent_orm).primary_key
+    matcher = NodeMatcher(graph)
+    return matcher.match(extract_name(parent_orm.__tablename__),
+                         **{pk.name: row[pk.name]}).first()
+
+
+def extract_name(tablename):
+    """Convert a Cordis table name to it's Neo4j Node label"""
```

Comment:
Please specify argument data type

Reply:
in this particular case the type is implicit since a) `tablename` is clearly a name, so a `str` b) only `str` have a `replace`, indexing and `title` method. Wrapping this all in a function is purely for readability.

However, the standard in python for functions which are not intended for external use is to prefix them with an underscore `_`, which is normally a get-out clause for doing proper doc strings - so I've done that now!

Thanks for pointing it out!